### PR TITLE
Coding style

### PR DESCRIPTION
**Pull request by janl**
- extract coding styles into separate document
- documented recommendations for using lodash on server and client

### CODE REVIEW

File: CODING_STYLE.md [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+```javascript
+var merge = require('lodash/merge')
+```
+
+If you want to use more than one function within one module, or if you want to combine multiple functions for a single operation, require the full lodash module:
+
+```javascript
+var _ = require('lodash')
+```
```
Comment (gr2m) [inline CODING_STYLE.md:49]: If I got it right, we don’t ever want to require the full lodash module for the client, we always want to do `require('lodash/{method}')`

Comment (janl) [inline CODING_STYLE.md:49]: hm, even if you use more than one?

``` javascript
var merge = require('lodash/merge')
var pluck = require('lodash/pluck')

merge(pluck(a), b)
```

File: CODING_STYLE.md [hunk 2]
```diff
@@ -0,0 +1,73 @@
+# Hoodie Coding Style Guidelines
+
+Please see [CONTRIBUTING.md](CONTRIBUTING.md) for more guidelines on
+contributing to Hoodie
```
Comment (varjmes) [inline CODING_STYLE.md]: Full stop at the end of the sentence.

Comment (janl) [inline CODING_STYLE.md]: fixed

**Comment by NickColley** [comment]
At some point do we want to add HTML, CSS? Or does that belong somewhere else.

**Comment by janl** [comment]
should definitely go here!

**Comment by NickColley** [comment]
In that case I'd like to propose http://codeguide.co/#html and http://codeguide.co/#css as a starting point, too bad there isn't a 'standard' for html and css.

File: CODING_STYLE.md [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+Prefer [loadash](https://lodash.com) over [Underscore](http://underscorejs.org "Underscore.js").
+
+
+## Client
+
+### Libraries with sub-modules that can be required individually, like lodash
+
+For client-side JavaScript code, it is important to limit the amount of code
+that is downloaded to the client to the code that is actually needed. The
+[loadash](https://lodash.com) library is a collection of utilities that are
```
Comment (Guria) [inline CODING_STYLE.md:33]: lodash

File: CODING_STYLE.md [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+- *Client*: coding styles that are only applicable to in-browser code.
+- *Server*: coding styles that are only applicable in server code.
+
+*Note: Client and Server coding styles can be contradicting, make sure to read
+these carefully*.
+
+
+## General
+
+Prefer [loadash](https://lodash.com) over [Underscore](http://underscorejs.org "Underscore.js").
```
Comment (Guria) [inline CODING_STYLE.md]: loadash -> lodash

File: CODING_STYLE.md (hunk 3 above)
Comment (janl) [inline CODING_STYLE.md:33]: fixed thanks!

**Comment by janl** [comment]
@nickcolley sure, wanna add these on top, or open a branch/pr once this is merged?

**Comment by NickColley** [comment]
@janl let's raise an issue so I don't block this. :)

**Comment by varjmes** [comment]
LGTM

**Comment by janl** [comment]
@nickcolley thanks, can you mention the issue here or this one there?

Also, one more lgtm pls ;)

File: CODING_STYLE.md [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+- *Client*: coding styles that are only applicable to in-browser code.
+- *Server*: coding styles that are only applicable in server code.
+
+*Note: Client and Server coding styles can be contradicting, make sure to read
+these carefully*.
+
+
+## General
+
+Prefer [lodash](https://lodash.com) over [Underscore](http://underscorejs.org "Underscore.js").
```
Comment (gr2m) [inline CODING_STYLE.md:24]: I don’t think this is worth mentioning any more. The [lodash / underscore merge](https://github.com/jashkenas/underscore/issues/2182) kinda sorta happened with lodash v4

**Comment by gr2m** [comment]
We could mention that in client code, we always use promises for async code and do promise based APIs. On server side we prefer callbacks unless it makes things more complicated

File: CODING_STYLE.md (hunk 5 above)
Comment (janl) [inline CODING_STYLE.md:24]: v4 _just_ came out, so I think it’s still good to have this in writing.

**Comment by janl** [comment]
> We could mention that in client code, we always use promises for async code and do promise based APIs. On server side we prefer callbacks unless it makes things more complicated

Yes, please push to this branch or open a PR once this lands :)

**Pull request closed without merging**